ports/esp32s2/common-hal/countio/Counter.c

```diff
@@ -0,0 +1,77 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 microDev
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "common-hal/countio/Counter.h"
+#include "common-hal/microcontroller/Pin.h"
+
+void common_hal_countio_counter_construct(countio_counter_obj_t* self,
+    const mcu_pin_obj_t* pin) {
+    claim_pin(pin);
+
+    // Prepare configuration for the PCNT unit
+    pcnt_config_t pcnt_config = {
+        // Set PCNT input signal and control GPIOs
+        .pulse_gpio_num = pin->number,
+        .ctrl_gpio_num = PCNT_PIN_NOT_USED,
+        .channel = PCNT_CHANNEL_0,
+        // What to do on the positive / negative edge of pulse input?
+        .pos_mode = PCNT_COUNT_INC,   // Count up on the positive edge
+        .neg_mode = PCNT_COUNT_DIS,   // Keep the counter value on the negative edge
+    };
+    // Initialize PCNT unit
+    pcnt_handler_init(&pcnt_config);
+
+    self->pin = pin->number;
+    self->unit = pcnt_config.unit;
+}
+
+bool common_hal_countio_counter_deinited(countio_counter_obj_t* self) {
+    return self->unit == PCNT_UNIT_MAX;
+}
+
+void common_hal_countio_counter_deinit(countio_counter_obj_t* self) {
+    if (common_hal_countio_counter_deinited(self)) {
+        return;
+    }
+    reset_pin_number(self->pin);
```

Comment:
This isn't a change request since I'm working on the fix in a new PR, but note that until we add in proper pin resets to Microcontroller/Pin.c, this function and the PCNT deinit might not actually release this pin from the PCNT. If you run into weird bugs after dynamically remapping pins, that's why.

Reply:
@hierophect I would love to collaborate with you on implementing proper pin resets.